README.md
![Clay Containers for implementing beautiful, modern neumorphic designs.](https://res.cloudinary.com/mca62511/image/upload/v1579847668/banner_zx6prd.png)

# Clay Containers

See the details on https://github.com/mcaubrey/clay_containers

# Clay Containers Plus

Create animation with implicit animated widgets.
This package simply change the `Container` to `AnimatedContainer`,
and wrap the `Text` with `AnimatedDefaultTextStyle`.


CHANGELOG.md
# v0.1.1+2 - Minor Fix

# v0.1.1+1 - Minor Fix

- Fix minor issues
- Apply `??` operator

# v0.1.0 - First Version

First version of clay containers plus.

- Animated Clay Container
- Animated Clay Text


example/main.dart
import 'package:clay_containers_plus/clay_containers.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with SingleTickerProviderStateMixin {
  Color baseColor = Color(0xFFf2f2f2);
  double firstDepth = 50;
  double secondDepth = 50;
  double thirdDepth = 50;
  double fourthDepth = 50;
  AnimationController _animationController;

  @override
  void initState() {
    _animationController = AnimationController(
      vsync: this, // the SingleTickerProviderStateMixin
      duration: Duration(seconds: 5),
    )..addListener(() {
        setState(() {});
      });

    _animationController.forward();

    super.initState();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    double stagger(value, progress, delay) {
      progress = progress - (1 - delay);
      if (progress < 0) progress = 0;
      return value * (progress / delay);
    }

    double calculatedFirstDepth =
        stagger(firstDepth, _animationController.value, 0.25);
    double calculatedSecondDepth =
        stagger(secondDepth, _animationController.value, 0.5);
    double calculatedThirdDepth =
        stagger(thirdDepth, _animationController.value, 0.75);
    double calculatedFourthDepth =
        stagger(fourthDepth, _animationController.value, 1);

    return Container(
      color: baseColor,
      child: Center(
        child: ClayContainer(
          color: baseColor,
          height: 240,
          width: 240,
          borderRadius: 200,
          curveType: CurveType.concave,
          spread: 30,
          depth: calculatedFirstDepth.toInt(),
          child: Center(
            child: ClayContainer(
              height: 200,
              width: 200,
              borderRadius: 200,
              depth: calculatedSecondDepth.toInt(),
              curveType: CurveType.convex,
              color: baseColor,
              child: Center(
                child: ClayContainer(
                    height: 160,
                    width: 160,
                    borderRadius: 200,
                    color: baseColor,
                    depth: calculatedThirdDepth.toInt(),
                    curveType: CurveType.concave,
                    child: Center(
                        child: ClayContainer(
                      height: 120,
                      width: 120,
                      borderRadius: 200,
                      color: baseColor,
                      depth: calculatedFourthDepth.toInt(),
                      curveType: CurveType.convex,
                    ))),
              ),
            ),
          ),
        ),
      ),
    );
  }
}


lib/widgets/clay_text.dart
import 'dart:math';

import 'package:flutter/material.dart';

class ClayText extends StatelessWidget {
  final String text;
  final Color color;
  final Color parentColor;
  final Color textColor;
  final TextStyle style;
  final double spread;
  final int depth;
  final double size;
  final bool emboss;
  final Duration duration;
  final Curve curve;
  final Function onEnd;
  final TextAlign textAlign;
  final bool softWrap;
  final TextOverflow overflow;
  final int maxLines;
  final TextWidthBasis textWidthBasis;
  final TextHeightBehavior textHeightBehavior;

  ClayText(
    this.text, {
    this.parentColor,
    this.textColor,
    this.color,
    this.spread,
    this.depth,
    this.style,
    this.size,
    this.emboss,
    this.duration = const Duration(milliseconds: 100),
    this.curve = Curves.linear,
    this.onEnd,
    this.maxLines,
    this.overflow = TextOverflow.clip,
    this.softWrap = true,
    this.textAlign,
    this.textHeightBehavior,
    this.textWidthBasis = TextWidthBasis.parent,
    Key key,
  }) : super(key: key);

  Color _getAdjustColor(Color baseColor, amount) {
    Map colors = {
      "red": baseColor.red,
      "green": baseColor.green,
      "blue": baseColor.blue
    };
    colors = colors.map((key, value) {
      if (value + amount < 0) return MapEntry(key, 0);
      if (value + amount > 255) return MapEntry(key, 255);
      return MapEntry(key, value + amount);
    });
    return Color.fromRGBO(colors["red"], colors["green"], colors["blue"], 1);
  }

  double _getSpread(base) {
    double calculated = (base / 10).floor().toDouble();
    return calculated == 0 ? 1 : calculated;
  }

  @override
  Widget build(BuildContext context) {
    final int depthValue = depth ?? 40;
    Color colorValue = color ?? Color(0xFFf0f0f0);
    final Color outerColorValue = parentColor ?? colorValue;
    double fontSizeValue = style?.fontSize ?? (size ?? 14);
    final TextStyle styleValue = style ?? DefaultTextStyle.of(context).style;
    final double spreadValue = spread ?? _getSpread(fontSizeValue);
    final bool embossValue = emboss ?? false;

    List<Shadow> shadowList = [
      Shadow(
          color: _getAdjustColor(
              outerColorValue, embossValue ? 0 - depthValue : depthValue),
          offset: Offset(0 - spreadValue / 2, 0 - spreadValue / 2),
          blurRadius: spreadValue),
      Shadow(
          color: _getAdjustColor(
              outerColorValue, embossValue ? depthValue : 0 - depthValue),
          offset: Offset(spreadValue / 2, spreadValue / 2),
          blurRadius: spreadValue)
    ];

    if (embossValue) shadowList = shadowList.reversed.toList();
    if (embossValue) colorValue = _getAdjustColor(colorValue, 0 - depthValue);
    if (textColor != null) colorValue = textColor;

    return AnimatedDefaultTextStyle(
      duration: duration,
      curve: curve,
      onEnd: onEnd,
      maxLines: maxLines,
      overflow: overflow,
      softWrap: softWrap,
      textAlign: textAlign,
      textHeightBehavior: textHeightBehavior,
      textWidthBasis: textWidthBasis,
      style: styleValue.copyWith(
        color: colorValue,
        shadows: shadowList,
        fontSize: fontSizeValue,
      ),
      child: Text(text),
    );
  }
}


lib/widgets/clay_containers.dart
import 'package:flutter/material.dart';

import '../constants.dart';

class ClayContainer extends StatelessWidget {
  final double height;
  final double width;
  final Color color;
  final Color parentColor;
  final Color surfaceColor;
  final double spread;
  final Widget child;
  final double borderRadius;
  final BorderRadius customBorderRadius;
  final CurveType curveType;
  final int depth;
  final bool emboss;
  final Curve curve;
  final Duration duration;
  final AlignmentGeometry alignment;
  final BoxConstraints constraints;
  final Decoration foregroundDecoration;
  final EdgeInsetsGeometry margin;
  final EdgeInsetsGeometry padding;
  final Matrix4 transform;
  final Function onEnd;

  ClayContainer({
    this.child,
    this.height,
    this.width,
    this.color,
    this.surfaceColor,
    this.parentColor,
    this.spread,
    this.borderRadius,
    this.customBorderRadius,
    this.curveType,
    this.depth,
    this.emboss,
    this.curve = Curves.linear,
    this.duration = const Duration(milliseconds: 100),
    this.alignment,
    this.constraints,
    this.foregroundDecoration,
    this.margin,
    this.padding,
    this.onEnd,
    this.transform,
    Key key,
  }) : super(key: key);

  Color _getAdjustColor(Color baseColor, amount) {
    Map colors = {
      "red": baseColor.red,
      "green": baseColor.green,
      "blue": baseColor.blue
    };
    colors = colors.map((key, value) {
      if (value + amount < 0) return MapEntry(key, 0);
      if (value + amount > 255) return MapEntry(key, 255);
      return MapEntry(key, value + amount);
    });
    return Color.fromRGBO(colors["red"], colors["green"], colors["blue"], 1);
  }

  List<Color> _getFlatGradients(baseColor, depth) {
    return [
      baseColor,
      baseColor,
    ];
  }

  List<Color> _getConcaveGradients(baseColor, depth) {
    return [
      _getAdjustColor(baseColor, 0 - depth),
      _getAdjustColor(baseColor, depth),
    ];
  }

  List<Color> _getConvexGradients(baseColor, depth) {
    return [
      _getAdjustColor(baseColor, depth),
      _getAdjustColor(baseColor, 0 - depth),
    ];
  }

  @override
  Widget build(BuildContext context) {
    final double heightValue = height;
    final double widthValue = width;
    final int depthValue = depth ?? 20;
    Color colorValue = color ?? Color(0xFFf0f0f0);
    final Color parentColorValue = parentColor ?? colorValue;
    final Color surfaceColorValue = surfaceColor ?? colorValue;
    final double spreadValue = spread ?? 6;
    final bool embossValue = emboss ?? false;
    BorderRadius borderRadiusValue =
        BorderRadius.circular(borderRadius ?? Radius.zero);

    if (customBorderRadius != null) {
      borderRadiusValue = customBorderRadius;
    }
    final CurveType curveTypeValue = curveType ?? CurveType.none;

    List<BoxShadow> shadowList = [
      BoxShadow(
          color: _getAdjustColor(
              parentColorValue, embossValue ? 0 - depthValue : depthValue),
          offset: Offset(0 - spreadValue, 0 - spreadValue),
          blurRadius: spreadValue),
      BoxShadow(
          color: _getAdjustColor(
              parentColorValue, embossValue ? depthValue : 0 - depthValue),
          offset: Offset(spreadValue, spreadValue),
          blurRadius: spreadValue)
    ];

    if (embossValue) shadowList = shadowList.reversed.toList();
    if (embossValue)
      colorValue = _getAdjustColor(colorValue, 0 - depthValue ~/ 2);
    if (surfaceColor != null) colorValue = surfaceColorValue;

    List<Color> gradientColors;
    switch (curveTypeValue) {
      case CurveType.concave:
        gradientColors = _getConcaveGradients(colorValue, depthValue);
        break;
      case CurveType.convex:
        gradientColors = _getConvexGradients(colorValue, depthValue);
        break;
      case CurveType.none:
        gradientColors = _getFlatGradients(colorValue, depthValue);
        break;
    }

    return AnimatedContainer(
      duration: duration,
      curve: curve,
      height: heightValue,
      width: widthValue,
      child: child,
      margin: margin,
      padding: padding,
      alignment: alignment,
      constraints: constraints,
      foregroundDecoration: foregroundDecoration,
      onEnd: onEnd,
      transform: transform,
      decoration: BoxDecoration(
          borderRadius: borderRadiusValue,
          color: colorValue,
          gradient: LinearGradient(
              begin: Alignment.topLeft,
              end: Alignment.bottomRight,
              colors: gradientColors),
          boxShadow: shadowList),
    );
  }
}
